### PR TITLE
Added SensorModelFactory class.  DO NOT MERGE.

### PR DESCRIPTION
**Pull request by twilson271828**
I haven't finished the SensorModelFactory::create method because I keep getting linker errors, so I've been working on the CMakeLists.txt file.  This is where I need help.

### CODE REVIEW

**Comment by twilson271828** [comment]
I corrected the CMakeLists.txt file so that the sensorutils lib links against the csm library as well as the CSM-CameraModel library.  I hard-coded the paths because I'm not certain what the final directory structure will look like.   Also, cmake kept finding an older version the csmapi lib that was in /usr/lib64.  When I deleted that, it found my local version on scratch.  The only undefined reference I am getting at present is to dlopen and dlerror.

**Comment by twilson271828** [comment]
Fixed the undefined ref linker error to dlopen.

File: CMakeLists.txt [hunk 1]
```diff
@@ -9,11 +9,28 @@ set(CMAKE_CXX_STANDARD 11)
 
 set(COVERAGE OFF CACHE BOOL "Coverage")
 
+
+#add_subdirectory(CSM-CameraModel)
+set(CAMERA_MODEL_INCLUDE_DIR,../CSM-CameraModel/include/usgscsm/)
+set(CAMERA_MODEL_LIB_DIR,../CSM-CameraModel/)
+set(CSMAPI_INCLUDE_DIR,../CSM-CameraModel/csm/include/csm/)
+set(CSMAPI_LIB_DIR,../CSM-CameraModel/csm/)
```
Comment (jessemapel) [inline CMakeLists.txt]: Why are these hardcoded to something outside of the project? The find_library call below should be the only thing setting these.

File: CMakeLists.txt [hunk 2]
```diff
@@ -9,11 +9,28 @@ set(CMAKE_CXX_STANDARD 11)
...
+
+
+find_library(CAMERA_MODEL_LIBRARY usgscsm PATHS ../CSM-CameraModel/)
```
Comment (jessemapel) [inline CMakeLists.txt]: The USGS CSM implementation library should not be linked into this library. The SET should be dynamically loading the implementation library at run time.

File: CMakeLists.txt [hunk 3]
```diff
@@ -43,10 +64,12 @@ target_include_directories(sensorutils
                            ${ARMADILLO_INCLUDE_DIRS}
 )
 
+message(${ARMADILLO_LIBRARIES})
 
 target_link_libraries(sensorutils PUBLIC
                       gtest ${CMAKE_THREAD_LIBS_INIT}
-                      ${ARMADILLO_LIBRARIES})
+                      ${ARMADILLO_LIBRARIES} ${CAMERA_MODEL_LIBRARY}
```
Comment (jessemapel) [inline CMakeLists.txt]: As stated above, the CAMERA_MODEL_LIBRARY should not be linked against. It is a plugin that should be loaded at run time.

File: CMakeLists.txt [hunk 4]
```diff
@@ -43,10 +64,12 @@ target_include_directories(sensorutils
...
+                       ${CSMAPI_LIBRARY} ${CMAKE_DL_LIBS})
```
Comment (jessemapel) [inline CMakeLists.txt]: I thought the SET was loading the plugin libraries, why is CMAKE_DL_LIBS being linked into this library?

File: include/sensormodel/SensorModelFactory.h [hunk 5]
```diff
@@ -0,0 +1,41 @@
+#ifndef SensorModelFactory_h
+#define SensorModelFactory_h
+
+#include "Plugin.h"
```
Comment (jessemapel) [inline include/sensormodel/SensorModelFactory.h]: Plugin.h is specific to CSM so it should be included in the implementation, not the header file

File: src/sensormodel/SensorModelFactory.cpp [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+ unique_ptr<SensorModel>  SensorModelFactory::create(const string &pathToPlugin,
+                                                     const string& pluginName ) {
+   string plugin = pathToPlugin +"/"+pluginName+".so";
+   void *pluginLib = dlopen(plugin.c_str(), RTLD_LAZY);
+   if (pluginLib == NULL) {
+     cerr << "There was an issue loading the plugin." << endl;
+     cerr << "\t" << dlerror() << endl;
+     return unique_ptr<SensorModel>(nullptr);
+   }
+   else {
```
Comment (jessemapel) [inline src/sensormodel/SensorModelFactory.cpp]: This entire block of code is the SET's job

**Comment by makaylas** [comment]
I'm working on the cmake stuff

**Comment by AstroKEW78** [comment]
Is our design document correct and up to date? Including the activity diagrams?
If not, this could be part of the confusion here...

**Comment by twilson271828** [comment]
I pushed a refactored version of the create method.  I have some design questions for the group, though.

- Are CSMSensorModel,IsisSensorModel supposed to represent UsgsAstroFrameSensorModel and UsgsAstroLsSensorModel or are these new classes?  If they're not, then we need to refactor them 
so that they both inherit from SensorModel which defines their common interface.  In this PR, I wasn't sure so I created a dummy class called IsisSensorModel.h
- There seems to be confusion about what are and are not the responsibilities of the SET.  If the SET instantiates the appropriate camera model from the ISD (that was my understanding as well), then what if the SensorModelFactory returns the appropriate plug-in instead of a pointer to a sensor model? (as it shows in the design).
-Where are the state data classes (UsgsAstroLsStateData...etc...) fitting into the design we have at present?

**Comment by jessemapel** [comment]
Here's my current understanding of our design as I won't be in for any discussion tomorrow:

* The CSMSensorModel and IsisSensorModel classes are wrappers around generic CSM and ISIS camera models respectively. The CSMSensorModel class should interface with the generic CSM api, not our implementation. The IsisSensorModel class should interface with the Isis Camera class, that is for later though.

* Here's how I see the responsibilities being divided.
* * SET: get the ISD from somewhere, convert it to a string, generate the metadata string, get the sensor model name, pass all of that to Sensor constructor.
* * Sensor: Take the meta data string and the sensor model name, parse the shape information out of the meta data string and pass it to the ShapeModelFactory, parse the sensor information out of the meta data string and pass that plus the sensor model name to the SensorModelFactory.
* * SensorModelFactory: parse the meta data and sensor model name to determine what type of SensorModel to create, if CSM get the appropriate plugin and ask it for the appropriate sensor model then create a CSMSensorModel object, if ISIS pass the cube label to the CameraModelFactory then create an IsisSensorModel object, return a SensorModel pointer to the object.

* The activity diagrams are old and have not been updated.

**Comment by jlaura** [comment]
@jessemapel :+1: Super succinct and spot on with what my understanding is as well.

**Comment by twilson271828** [comment]
Thank you everyone who helped get this in!

**Pull request closed without merging**